INTRODUCTION

"Would you recommend a location in Toronto or New York City to open a new Peruvian restaurant?" 
The increasing popularity of Peruvian food has been unstoppable since 2011 and the Culinary Institute of America named "the Year of Peruvian Cuisine" in 2014. 
Therefore, my partner, who is a Peruvian entrepreneur and a passionate of food, wants to open a new Peruvian restaurant. He explains to me that the new restaurant should have shopping places nearby and transportation facilities within 5 minutes. He wants me concentrated on the selection of the restaurant's location according to its nearby environment. Restaurant facility and the rental price is not my concern. 

This project will explore the similarities and dissimilarities between specific neighborhoods in the two cities, and determine which areas best fit the Peruvian restaurant customers

In [3]:
import numpy as np 

import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json 

from geopy.geocoders import Nominatim 

import requests 
from pandas.io.json import json_normalize 

import matplotlib.cm as cm
import matplotlib.colors as colors


from sklearn.cluster import KMeans

print('Libraries imported.')

Libraries imported.


In [4]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium

print('Folium installed and imported!')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    folium-0.5.0               |             py_0          45 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    branca-0.4.1               |             py_0          26 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         713 KB

The following NEW packages will be INSTALLED:

    altair:  4.1.0-py_1 conda-forge
    branca:  0.4.1-py_0 conda-forge
    folium:  0.5.0-py_0 conda-forge
    vincent: 0.4.4-py_1 conda-forge


folium-0.5.0         | 45 KB     | ##################################### | 1

Use data from Toronto Neighborhood Wikipedia page to segment, cluster and explore neighborhoods in Toronto
The dataset being used is found at https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M.

In [8]:
table = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M', header = 0)

df_toronto = table[0]
df_toronto

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
33          M7E      Not assigned   
34          M8E      Not assigned   
35          M9E      Not assigned   
36          M1G       Scarborough   
37          M2G      Not assigned   
38          M3G      Not assigned   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
42          M7G      Not assigned   
43          M8G      Not assigned   
44          M9G      Not assigned   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
51          M7H      Not assigned   
52          M8H      Not assigned   
53          M9H      Not assigned   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
60          M7J      Not assigned   
61          M8J      Not assigned   
62          M9J      Not assigned   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
69          M7K      Not assigned   
70          M8K      Not assigned   
71          M9K      Not assigned   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
78          M7L      Not assigned   
79          M8L      Not assigned   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
87          M7M      Not assigned   
88          M8M      Not assigned   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
96          M7N      Not assigned   
97          M8N      Not assigned   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
101         M3P      Not assigned   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
105         M7P      Not assigned   
106         M8P      Not assigned   
107     

In [9]:
df_toronto.shape

(180, 3)

In [10]:
df_toronto.to_csv('Toronto.TASK_1_df.csv',index=False)

In [19]:
task1_data_csv = "Toronto.TASK_1_df.csv"
df_toronto = pd.read_csv(task1_data_csv).set_index("Postal Code")
df_toronto.rename_axis("Postal Code", axis='index', inplace=True)
df_toronto.drop(df_toronto[df_toronto.Borough == 'Not assigned'].index, inplace=True)
df_toronto = df_toronto.groupby(['Postal Code', 'Borough'])['Neighborhood'].apply(lambda x: ','.join(x)).reset_index()
df_toronto.loc[85,'Neighborhood'] = 'Queen\'s Park'
print (df_toronto.shape)

df_toronto.head()

(103, 3)


Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

Complete the information neighborhood table with latitude and longitude information

In [20]:
latlong = pd.read_csv("http://cocl.us/Geospatial_data")
latlong.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [25]:
latlong.rename(columns = {"Postal Code": "PostalCode"}, inplace = True)
print(latlong.shape)

latlong.head()

(103, 3)


PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [24]:
latlong.tail()

PostalCode   Latitude  Longitude
98         M9N  43.706876 -79.518188
99         M9P  43.696319 -79.532242
100        M9R  43.688905 -79.554724
101        M9V  43.739416 -79.588437
102        M9W  43.706748 -79.594054

In [39]:
toronto_neighborhoods = df_toronto.join(latlong)
toronto_neighborhoods.head()
toronto_neighborhoods.to_csv('Toronto.TASK_II_df.csv',index=False)
toronto_neighborhoods.shape
toronto_neighborhoods

Postal Code           Borough  \
0           M1B       Scarborough   
1           M1C       Scarborough   
2           M1E       Scarborough   
3           M1G       Scarborough   
4           M1H       Scarborough   
5           M1J       Scarborough   
6           M1K       Scarborough   
7           M1L       Scarborough   
8           M1M       Scarborough   
9           M1N       Scarborough   
10          M1P       Scarborough   
11          M1R       Scarborough   
12          M1S       Scarborough   
13          M1T       Scarborough   
14          M1V       Scarborough   
15          M1W       Scarborough   
16          M1X       Scarborough   
17          M2H        North York   
18          M2J        North York   
19          M2K        North York   
20          M2L        North York   
21          M2M        North York   
22          M2N        North York   
23          M2P        North York   
24          M2R        North York   
25          M3A        North York   
26          M3B        North York   
27          M3C        North York   
28          M3H        North York   
29          M3J        North York   
30          M3K        North York   
31          M3L        North York   
32          M3M        North York   
33          M3N        North York   
34          M4A        North York   
35          M4B         East York   
36          M4C         East York   
37          M4E      East Toronto   
38          M4G         East York   
39          M4H         East York   
40          M4J         East York   
41          M4K      East Toronto   
42          M4L      East Toronto   
43          M4M      East Toronto   
44          M4N   Central Toronto   
45          M4P   Central Toronto   
46          M4R   Central Toronto   
47          M4S   Central Toronto   
48          M4T   Central Toronto   
49          M4V   Central Toronto   
50          M4W  Downtown Toronto   
51          M4X  Downtown Toronto   
52          M4Y  Downtown Toronto   
53          M5A  Downtown Toronto   
54          M5B  Downtown Toronto   
55          M5C  Downtown Toronto   
56          M5E  Downtown Toronto   
57          M5G  Downtown Toronto   
58          M5H  Downtown Toronto   
59          M5J  Downtown Toronto   
60          M5K  Downtown Toronto   
61          M5L  Downtown Toronto   
62          M5M        North York   
63          M5N   Central Toronto   
64          M5P   Central Toronto   
65          M5R   Central Toronto   
66          M5S  Downtown Toronto   
67          M5T  Downtown Toronto   
68          M5V  Downtown Toronto   
69          M5W  Downtown Toronto   
70          M5X  Downtown Toronto   
71          M6A        North York   
72          M6B        North York   
73          M6C              York   
74          M6E              York   
75          M6G  Downtown Toronto   
76          M6H      West Toronto   
77          M6J      West Toronto   
78          M6K      West Toronto   
79          M6L        North York   
80          M6M              York   
81          M6N              York   
82          M6P      West Toronto   
83          M6R      West Toronto   
84          M6S      West Toronto   
85          M7A  Downtown Toronto   
86          M7R       Mississauga   
87          M7Y      East Toronto   
88          M8V         Etobicoke   
89          M8W         Etobicoke   
90          M8X         Etobicoke   
91          M8Y         Etobicoke   
92          M8Z         Etobicoke   
93          M9A         Etobicoke   
94          M9B         Etobicoke   
95          M9C         Etobicoke   
96          M9L        North York   
97          M9M        North York   
98          M9N              York   
99          M9P         Etobicoke   
100         M9R         Etobicoke   
101         M9V         Etobicoke   
102         M9W         Etobicoke   

                                          Neighborhood PostalCode   Latitude  \
0                                       Malvern, Rouge        M1B  43.80668

In [42]:
print('Toronto has {} boroughs and {} neighborhoods.'.format(
        len(toronto_neighborhoods['Borough'].unique()),
        toronto_neighborhoods.shape[0]
    )
)

Toronto has 10 boroughs and 103 neighborhoods.


In [43]:
address = 'Toronto, CA'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto, Canada are {}, {}.'.format(latitude, longitude))

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.21.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


The geograpical coordinate of Toronto, Canada are 43.6534817, -79.3839347.


Create a map of Toronto with neighborhoods

In [46]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

for lat, lng, borough, neighborhood in zip(toronto_neighborhoods['Latitude'], toronto_neighborhoods['Longitude'], toronto_neighborhoods['Borough'], toronto_neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [139]:
Downtown_data = toronto_neighborhoods[toronto_neighborhoods['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
Downtown_data.head()


Postal Code           Borough                 Neighborhood PostalCode  \
0         M4W  Downtown Toronto                     Rosedale        M4W   
1         M4X  Downtown Toronto  St. James Town, Cabbagetown        M4X   
2         M4Y  Downtown Toronto         Church and Wellesley        M4Y   
3         M5A  Downtown Toronto    Regent Park, Harbourfront        M5A   
4         M5B  Downtown Toronto     Garden District, Ryerson        M5B   

    Latitude  Longitude  
0  43.679563 -79.377529  
1  43.667967 -79.367675  
2  43.665860 -79.383160  
3  43.654260 -79.360636  
4  43.657162 -79.378937

In [140]:
Downtown_data.shape

(19, 6)

Create a map of Downtown Toronto, Toronto, Canada

In [141]:
address = 'Downtown Toronto, Toronto'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto, CA are {}, {}.'.format(latitude, longitude))

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.21.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


The geograpical coordinate of Downtown Toronto, CA are 43.6541737, -79.38081164513409.


In [142]:
map_Downtown = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, lng, label in zip(Downtown_data['Latitude'], Downtown_data['Longitude'], Downtown_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Downtown)  
    
map_Downtown

Define Foursquare credentials and version

In [53]:
CLIENT_ID = 'XNBNBG20BYSDIUYVKSGBA53XFSOMDDIIKDRLT5RRTT05MKQ0'
CLIENT_SECRET = 'XIQULR4IFIK3SL22CYAUTXAGZ1OLIAVVHWPFEJMNPQNFAH21'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: XNBNBG20BYSDIUYVKSGBA53XFSOMDDIIKDRLT5RRTT05MKQ0
CLIENT_SECRET:XIQULR4IFIK3SL22CYAUTXAGZ1OLIAVVHWPFEJMNPQNFAH21


In [143]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        
            
        
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            100)
            
        
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    print('Found {} venues in {} neighborhoods.'.format(nearby_venues.shape[0], len(venues_list)))
    
    return(nearby_venues)

Analyze each Neighborhood

In [144]:
Downtown_venues = getNearbyVenues(names=Downtown_data['Neighborhood'],
                                   latitudes=Downtown_data['Latitude'],
                                   longitudes=Downtown_data['Longitude']
                                  )

Found 1228 venues in 19 neighborhoods.


In [146]:
print(Downtown_venues.shape)
Downtown_venues.head() 

(1228, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                     Rosedale              43.679563              -79.377529   
1                     Rosedale              43.679563              -79.377529   
2                     Rosedale              43.679563              -79.377529   
3                     Rosedale              43.679563              -79.377529   
4  St. James Town, Cabbagetown              43.667967              -79.367675   

                  Venue  Venue Latitude  Venue Longitude Venue Category  
0         Rosedale Park       43.682328       -79.378934     Playground  
1          Whitney Park       43.682036       -79.373788           Park  
2  Alex Murray Parkette       43.678300       -79.382773           Park  
3        Milkman's Lane       43.676352       -79.373842          Trail  
4           Cranberries       43.667843       -79.369407          Diner

In [147]:
Downtown_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     62   
Christie                                                               17   
Church and Wellesley                                                   78   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              56   
Queen's Park                                                           39   
Regent Park, Harbourfront                                              48   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
St. James Town                                                         77   
St. James Town, Cabbagetown                                            50   
Stn A PO Boxes                                                         95   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      62   
Christie                                                                17   
Church and Wellesley                                                    78   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               56   
Queen's Park                                                            39   
Regent Park, Harbourfront                                               48   
Richmond, Adelaide, King                                                94   
Rosedale                                                                 4   
St. James Town                                                          77   
St. James Town, Cabbagetown                                             50   
Stn A PO Boxes                                                          95   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          35   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            57              57   
CN Tower, King and Spadina, Railway Lands, Harb...     16              16   
Central Bay Street                                     62              62   
Christie                                               17              17   
Church and Wellesley                                   78              78   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

In [148]:
print('There are {} distinct venues in {} categories.'.format(
    len(Downtown_venues['Venue'].unique()),len(Downtown_venues['Venue Category'].unique())))

There are 717 distinct venues in 210 categories.


In [151]:
Downtown_onehot = pd.get_dummies(Downtown_venues[['Venue Category']], prefix="", prefix_sep="")


Downtown_onehot['Neighborhood'] = Downtown_venues['Neighborhood'] 


fixed_columns = [Downtown_onehot.columns[-1]] + list(Downtown_onehot.columns[:-1])
Downtown_onehot = Downtown_onehot[fixed_columns]

Downtown_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  \
0                 0                   0          0           0           0   
1                 0                   0          0           0           0   
2                 0                   0          0           0           0   
3                 0                   0          0           0           0   
4                 0                   0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       0     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Butcher  \
0        0                0         0             0              0        0   
1        0                0         0             0              0        0   
2        0                0         0             0              0        0   
3        0                0         0             0              0        0   
4        0                0         0             0              0        0   

   Café  Candy Store  Caribbean Restaurant  Cheese Shop  Chinese Restaurant  \
0     0            0                     0            0                   0   
1     0            0                     0            0                   0   
2     0            0                     0            0                   0   
3     0            0     

In [152]:
Downtown_grouped = Downtown_onehot.groupby('Neighborhood').mean().reset_index()
Downtown_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
2                                  Central Bay Street     0.016129   
3                                            Christie     0.000000   
4                                Church and Wellesley     0.025641   
5                      Commerce Court, Victoria Hotel     0.000000   
6              First Canadian Place, Underground city     0.000000   
7                            Garden District, Ryerson     0.000000   
8   Harbourfront East, Union Station, Toronto Islands     0.000000   
9           Kensington Market, Chinatown, Grange Park     0.000000   
10                                       Queen's Park     0.025641   
11                          Regent Park, Harbourfront     0.020833   
12                           Richmond, Adelaide, King     0.000000   
13                                           Rosedale     0.000000   
14                                     St. James Town     0.000000   
15                        St. James Town, Cabbagetown     0.000000   
16                                     Stn A PO Boxes     0.010526   
17           Toronto Dominion Centre, Design Exchange     0.000000   
18                     University of Toronto, Harbord     0.028571   

    Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0.000000   0.0000              0.0000        0.0000   
1            0.000000   0.0625              0.0625        0.0625   
2            0.000000   0.0000              0.0000        0.0000   
3            0.000000   0.0000              0.0000        0.0000   
4            0.012821   0.0000              0.0000        0.0000   
5            0.000000   0.0000              0.0000        0.0000   
6            0.000000   0.0000              0.0000        0.0000   
7            0.000000   0.0000              0.0000        0.0000   
8            0.000000   0.0000              0.0000        0.0000   
9            0.000000   0.0000              0.0000        0.0000   
10           0.000000   0.0000              0.0000        0.0000   
11           0.000000   0.0000              0.0000        0.0000   
12           0.000000   0.0000              0.0000        0.0000   
13           0.000000   0.0000              0.0000        0.0000   
14           0.000000   0.0000              0.0000        0.0000   
15           0.000000   0.0000              0.0000        0.0000   
16           0.000000   0.0000              0.0000        0.0000   
17           0.000000   0.0000              0.0000        0.0000   
18           0.000000   0.0000              0.0000        0.0000   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0            0.000            0.000             0.000             0.000000   
1            0.125            0.125             0.125             0.000000   
2            0.000            0.000             0.000             0.000000   
3            0.000            0.000             0.000             0.000000   
4            0.000            0.000             0.000             0.012821   
5            0.000            0.000             0.000             0.040000   
6            0.000            0.000             0.000             0.030000   
7            0.000            0.000             0.000             0.000000   
8            0.000            0.000             0.000             0.000000   
9            0.000            0.000             0.000             0.000000   
10           0.000            0.000             0.000             0.000000   
11           0.000            0.000             0.000             0.000000   
12           0.000            0.000             0.000             0.021277   
13           0.000            0.000             0.000             0.000000   
14           0.000            0.000             0.000             0.038961   
15           0.000            0

Define top venues

In [153]:
num_top_venues = 10

for hood in Downtown_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Downtown_grouped[Downtown_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2            Beer Bar  0.04
3  Seafood Restaurant  0.04
4                Café  0.04
5              Bakery  0.04
6         Cheese Shop  0.04
7          Restaurant  0.04
8        Concert Hall  0.02
9        Gourmet Shop  0.02


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                 venue  freq
0       Airport Lounge  0.12
1      Airport Service  0.12
2     Airport Terminal  0.12
3        Boat or Ferry  0.06
4      Harbor / Marina  0.06
5             Boutique  0.06
6     Sculpture Garden  0.06
7  Rental Car Location  0.06
8                Plane  0.06
9          Coffee Shop  0.06


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.18
1   Italian Restaurant  0.08
2                 Café  0.06
3       Sandwich Place  0.05
4      Bubble Tea Shop  0.03
5       Ice Cream Shop  0.03
6   

In [154]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [155]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Downtown_grouped['Neighborhood']

for ind in np.arange(Downtown_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Downtown_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1   CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
2                                  Central Bay Street           Coffee Shop   
3                                            Christie         Grocery Store   
4                                Church and Wellesley           Coffee Shop   
5                      Commerce Court, Victoria Hotel           Coffee Shop   
6              First Canadian Place, Underground city           Coffee Shop   
7                            Garden District, Ryerson        Clothing Store   
8   Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
9           Kensington Market, Chinatown, Grange Park                  Café   
10                                       Queen's Park           Coffee Shop   
11                          Regent Park, Harbourfront           Coffee Shop   
12                           Richmond, Adelaide, King           Coffee Shop   
13                                           Rosedale                  Park   
14                                     St. James Town                  Café   
15                        St. James Town, Cabbagetown           Coffee Shop   
16                                     Stn A PO Boxes           Coffee Shop   
17           Toronto Dominion Centre, Design Exchange           Coffee Shop   
18                     University of Toronto, Harbord                  Café   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0           Cocktail Bar            Restaurant               Beer Bar   
1        Airport Service      Airport Terminal       Sculpture Garden   
2     Italian Restaurant                  Café         Sandwich Place   
3                   Café                  Park            Coffee Shop   
4       Sushi Restaurant   Japanese Restaurant             Restaurant   
5             Restaurant                  Café                  Hotel   
6                   Café                   Gym             Restaurant   
7            Coffee Shop                  Café         Cosmetics Shop   
8               Aquarium                  Café                  Hotel   
9            Coffee Shop    Mexican Restaurant  Vietnamese Restaurant   
10      Sushi Restaurant                   Gym                  Diner   
11                Bakery                  Park                    Pub   
12                  Café            Restaurant                    Gym   
13                 Trail            Playground          Women's Store   
14           Coffee Shop   American Restaurant              Gastropub   
15            Restaurant                Bakery            Pizza Place   
16                  Café   Japanese Restaurant     Seafood Restaurant   
17                  Café                 Hotel             Restaurant   
18             Bookstore                   Bar     Italian Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                 Bakery    Seafood Restaurant                  Café   
1          Boat or Ferry               Airport    Airport Food Court   
2            Salad Place        Ice Cream Shop   Japanese Restaurant   
3             Baby Store    Athletics & Sports           Candy Store   
4            Yoga Studio            Smoke Shop               Gay Bar   
5                    Gym   American Restaurant    Seafood Restaurant   
6    Japanese Restaurant                 Hotel           Salad Place   
7             Restaurant    Italian Restaurant       Bubble Tea Shop   
8    Sporting Goods Shop   Fried Chicken Joint            Restaurant   
9                 Bakery                   Bar           Gaming Cafe   
10                  Park    Mexican Restaurant             Juice Bar   
11        Breakfast Spot            Restaurant               Theater   
12       Thai Restaurant                 Hotel        Clothing 

In [156]:
neighborhoods_venues_sorted.iloc[11,]

Neighborhood              Regent Park, Harbourfront
1st Most Common Venue                   Coffee Shop
2nd Most Common Venue                        Bakery
3rd Most Common Venue                          Park
4th Most Common Venue                           Pub
5th Most Common Venue                Breakfast Spot
6th Most Common Venue                    Restaurant
7th Most Common Venue                       Theater
8th Most Common Venue                          Café
9th Most Common Venue                          Bank
10th Most Common Venue          Distribution Center
Name: 11, dtype: object

Cluster Neighborhood

In [157]:
kclusters = 3

Downtown_grouped_clustering = Downtown_grouped.drop('Neighborhood', 1)


kmeans = KMeans(n_clusters=kclusters, random_state=2).fit(Downtown_grouped_clustering)

 
kmeans.labels_

array([0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0],
      dtype=int32)

In [158]:
Downtown_data.drop(Downtown_data[scarborough_data.Neighborhood == 'Upper Rouge'].index, inplace = True)

Downtown_merged = Downtown_data

Downtown_merged['Cluster Labels'] = kmeans.labels_


Downtown_merged = Downtown_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Downtown_merged.head() 

Postal Code           Borough                 Neighborhood PostalCode  \
0         M4W  Downtown Toronto                     Rosedale        M4W   
1         M4X  Downtown Toronto  St. James Town, Cabbagetown        M4X   
2         M4Y  Downtown Toronto         Church and Wellesley        M4Y   
3         M5A  Downtown Toronto    Regent Park, Harbourfront        M5A   
4         M5B  Downtown Toronto     Garden District, Ryerson        M5B   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.679563 -79.377529               0                  Park   
1  43.667967 -79.367675               0           Coffee Shop   
2  43.665860 -79.383160               0           Coffee Shop   
3  43.654260 -79.360636               1           Coffee Shop   
4  43.657162 -79.378937               0        Clothing Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                 Trail            Playground         Women's Store   
1            Restaurant                Bakery           Pizza Place   
2      Sushi Restaurant   Japanese Restaurant            Restaurant   
3                Bakery                  Park                   Pub   
4           Coffee Shop                  Café        Cosmetics Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Creperie      Doner Restaurant               Dog Run   
1                  Café                   Pub                  Park   
2           Yoga Studio            Smoke Shop               Gay Bar   
3        Breakfast Spot            Restaurant               Theater   
4            Restaurant    Italian Restaurant       Bubble Tea Shop   

  8th Most Common Venue      9th Most Common Venue    10th Most Common Venue  
0   Distribution Center             Discount Store                     Diner  
1    Italian Restaurant         Chinese Restaurant                 Pet Store  
2                 Hotel                  Gastropub  Mediterranean Restaurant  
3                  Café                       Bank       Distribution Center  
4   Japanese Restaurant  Middle Eastern Restaurant                   Theater

In [159]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)


x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]


markers_colors = []
for lat, lon, poi, cluster in zip(Downtown_merged['Latitude'], Downtown_merged['Longitude'], Downtown_merged['Neighborhood'], Downtown_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [163]:
Downtown_cluster_0 = Downtown_merged.loc[Downtown_merged['Cluster Labels'] == 0, Downtown_merged.columns[[1] + list(range(4, Downtown_merged.shape[1]))]]

Downtown_cluster_1 = Downtown_merged.loc[Downtown_merged['Cluster Labels'] == 1, Downtown_merged.columns[[1] + list(range(4, Downtown_merged.shape[1]))]]

Downtown_cluster_2 = Downtown_merged.loc[Downtown_merged['Cluster Labels'] == 2, Downtown_merged.columns[[1] + list(range(4, Downtown_merged.shape[1]))]]

In [164]:
Downtown_cluster_0

Borough   Latitude  Longitude  Cluster Labels  \
0   Downtown Toronto  43.679563 -79.377529               0   
1   Downtown Toronto  43.667967 -79.367675               0   
2   Downtown Toronto  43.665860 -79.383160               0   
4   Downtown Toronto  43.657162 -79.378937               0   
5   Downtown Toronto  43.651494 -79.375418               0   
6   Downtown Toronto  43.644771 -79.373306               0   
7   Downtown Toronto  43.657952 -79.387383               0   
8   Downtown Toronto  43.650571 -79.384568               0   
9   Downtown Toronto  43.640816 -79.381752               0   
10  Downtown Toronto  43.647177 -79.381576               0   
11  Downtown Toronto  43.648198 -79.379817               0   
12  Downtown Toronto  43.662696 -79.400049               0   
14  Downtown Toronto  43.628947 -79.394420               0   
15  Downtown Toronto  43.646435 -79.374846               0   
16  Downtown Toronto  43.648429 -79.382280               0   
17  Downtown Toronto  43.669542 -79.422564               0   
18  Downtown Toronto  43.662301 -79.389494               0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                   Park                 Trail            Playground   
1            Coffee Shop            Restaurant                Bakery   
2            Coffee Shop      Sushi Restaurant   Japanese Restaurant   
4         Clothing Store           Coffee Shop                  Café   
5                   Café           Coffee Shop   American Restaurant   
6            Coffee Shop          Cocktail Bar            Restaurant   
7            Coffee Shop    Italian Restaurant                  Café   
8            Coffee Shop                  Café            Restaurant   
9            Coffee Shop              Aquarium                  Café   
10           Coffee Shop                  Café                 Hotel   
11           Coffee Shop            Restaurant                  Café   
12                  Café             Bookstore                   Bar   
14        Airport Lounge       Airport Service      Airport Terminal   
15           Coffee Shop                  Café   Japanese Restaurant   
16           Coffee Shop                  Café                   Gym   
17         Grocery Store                  Café                  Park   
18           Coffee Shop      Sushi Restaurant                   Gym   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0          Women's Store              Creperie      Doner Restaurant   
1            Pizza Place                  Café                   Pub   
2             Restaurant           Yoga Studio            Smoke Shop   
4         Cosmetics Shop            Restaurant    Italian Restaurant   
5              Gastropub          Cocktail Bar                   Gym   
6               Beer Bar                Bakery    Seafood Restaurant   
7         Sandwich Place           Salad Place        Ice Cream Shop   
8                    Gym       Thai Restaurant                 Hotel   
9                  Hotel   Sporting Goods Shop   Fried Chicken Joint   
10            Restaurant   Japanese Restaurant           Salad Place   
11                 Hotel                   Gym   American Restaurant   
12    Italian Restaurant   Japanese Restaurant                Bakery   
14      Sculpture Garden         Boat or Ferry               Airport   
15    Seafood Restaurant            Restaurant          Cocktail Bar   
16            Restaurant   Japanese Restaurant                 Hotel   
17           Coffee Shop            Baby Store    Athletics & Sports   
18                 Diner                  Park    Mexican Restaurant   

   7th Most Common Venue 8th Most Common Venue      9th Most Common Venue  \
0                Dog Run   Distribution Center             Discount Store   
1                   Park    Italian Restaurant         Chinese Restaurant   
2                Gay Bar                 Hotel                  Gastrop

In [165]:
Downtown_cluster_1

Borough  Latitude  Longitude  Cluster Labels  \
3  Downtown Toronto  43.65426 -79.360636               1   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
3           Coffee Shop                Bakery                  Park   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
3                   Pub        Breakfast Spot            Restaurant   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
3               Theater                  Café                  Bank   

  10th Most Common Venue  
3    Distribution Center

In [167]:
Downtown_cluster_2

Borough   Latitude  Longitude  Cluster Labels  \
13  Downtown Toronto  43.653206 -79.400049               2   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
13                  Café           Coffee Shop    Mexican Restaurant   

    4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
13  Vietnamese Restaurant                Bakery                   Bar   

   7th Most Common Venue          8th Most Common Venue 9th Most Common Venue  \
13           Gaming Cafe  Vegetarian / Vegan Restaurant          Dessert Shop   

   10th Most Common Venue  
13          Burrito Place

Use data from New York Neighborhood to segment, cluster and explore neighborhoods in NY
The dataset being used is found at https://geo.nyu.edu/catalog/nyu_2451_34572

In [169]:
!wget -q -O 'newyork_data.json' https://ibm.box.com/shared/static/fbpwbovar7lf8p5sgddm06cgipa2rxpe.json
print('Data downloaded!')

Data downloaded!


In [170]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [171]:
neighborhoods_data = newyork_data['features']

In [172]:
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

In [78]:
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

neighborhoods = pd.DataFrame(columns=column_names)

Complete the information neighborhood table with latitude and longitude information

In [173]:
 for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [174]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [175]:
neighborhoods.shape

(612, 4)

Analyze each Neighborhood

In [176]:

print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 612 neighborhoods.


In [177]:

address = 'New York City, NY'

geolocator = Nominatim(user_agent = 'my-application')
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


Create a map of NT with neighborhoods

In [178]:

map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

In [210]:
queens_data = neighborhoods[neighborhoods['Borough'] == 'Queens'].reset_index(drop=True)
queens_data.head()

Borough     Neighborhood   Latitude  Longitude
0  Queens          Astoria  40.768509 -73.915654
1  Queens         Woodside  40.746349 -73.901842
2  Queens  Jackson Heights  40.751981 -73.882821
3  Queens         Elmhurst  40.744049 -73.881656
4  Queens     Howard Beach  40.654225 -73.838138

In [211]:
address = 'Queens, NY'

geolocator = Nominatim(user_agent = 'my-application')
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Queens are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Queens are 40.7498243, -73.7976337.


In [212]:
map_queens = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, lng, label in zip(queens_data['Latitude'], queens_data['Longitude'], queens_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_queens)  
    
map_queens

In [234]:
queens_data.loc[10, 'Neighborhood']

'Long Island City'

In [235]:
neighborhood_latitude = queens_data.loc[10, 'Latitude'] 
neighborhood_longitude = queens_data.loc[10, 'Longitude']

neighborhood_name = queens_data.loc[10, 'Neighborhood']

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Long Island City are 40.75021734610528, -73.93920223915505.


In [236]:
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

url

'https://api.foursquare.com/v2/venues/explore?&client_id=XNBNBG20BYSDIUYVKSGBA53XFSOMDDIIKDRLT5RRTT05MKQ0&client_secret=XIQULR4IFIK3SL22CYAUTXAGZ1OLIAVVHWPFEJMNPQNFAH21&v=20180605&ll=40.75021734610528,-73.93920223915505&radius=500&limit=100'

In [237]:
results = requests.get(url).json()

In [238]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [239]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) 


filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]


nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)


nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name    categories        lat        lng
0                            Chip  Dessert Shop  40.750069 -73.940831
1               Etto Espresso Bar   Coffee Shop  40.748703 -73.940689
2               Hilton Garden Inn         Hotel  40.750216 -73.936886
3                     Dutch Kills  Cocktail Bar  40.747830 -73.940108
4  Brooklyn Boulders Queensbridge  Climbing Gym  40.752649 -73.940010

In [240]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

71 venues were returned by Foursquare.


In [241]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        #print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    print('Found {} venues in {} neighborhoods.'.format(nearby_venues.shape[0], len(venues_list)))
    
    return(nearby_venues)

Define top venues

In [242]:
queens_venues = getNearbyVenues(names=queens_data['Neighborhood'],
                                   latitudes=queens_data['Latitude'],
                                   longitudes=queens_data['Longitude'])

Found 4160 venues in 162 neighborhoods.


In [243]:
queens_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0      Astoria              40.768509              -73.915654   
1      Astoria              40.768509              -73.915654   
2      Astoria              40.768509              -73.915654   
3      Astoria              40.768509              -73.915654   
4      Astoria              40.768509              -73.915654   

              Venue  Venue Latitude  Venue Longitude        Venue Category  
0      Favela Grill       40.767348       -73.917897  Brazilian Restaurant  
1    Orange Blossom       40.769856       -73.917012          Gourmet Shop  
2  Titan Foods Inc.       40.769198       -73.919253          Gourmet Shop  
3   CrossFit Queens       40.769404       -73.918977                   Gym  
4      Off The Hook       40.767200       -73.918104    Seafood Restaurant

In [244]:
print(queens_venues.shape)
queens_venues.tail()

(4160, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
4155  Queensbridge              40.756091              -73.945631   
4156  Queensbridge              40.756091              -73.945631   
4157  Queensbridge              40.756091              -73.945631   
4158  Queensbridge              40.756091              -73.945631   
4159  Queensbridge              40.756091              -73.945631   

                               Venue  Venue Latitude  Venue Longitude  \
4155  Queensbridge Basketball Courts       40.755060       -73.949103   
4156             The Ravel Hotel Gym       40.753787       -73.948815   
4157              Profundo Pool Club       40.753719       -73.948878   
4158         Estate Garden And Grill       40.753700       -73.948841   
4159                       Track 114       40.753008       -73.947833   

          Venue Category  
4155    Basketball Court  
4156  Athletics & Sports  
4157          Hotel Pool  
4158         Beer Garden  
4159            Platform

In [245]:
queens_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                 
Arverne                                  36                      36     36   
Astoria                                 196                     196    196   
Astoria Heights                          26                      26     26   
Auburndale                               36                      36     36   
Bay Terrace                              76                      76     76   
Bayside                                 146                     146    146   
Bayswater                                 4                       4      4   
Beechhurst                               30                      30     30   
Bellaire                                 26                      26     26   
Belle Harbor                             34                      34     34   
Bellerose                                38                      38     38   
Blissville                               34                      34     34   
Breezy Point                              6                       6      6   
Briarwood                                22                      22     22   
Broad Channel                            12                      12     12   
Brookville                                2                       2      2   
Cambria Heights                          28                      28     28   
College Point                            70                      70     70   
Corona                                   42                      42     42   
Douglaston                               46                      46     46   
East Elmhurst                            22                      22     22   
Edgemere                                 28                      28     28   
Elmhurst                                 48                      48     48   
Far Rockaway                             60                      60     60   
Floral Park                              12                      12     12   
Flushing                                124                     124    124   
Forest Hills                             74                      74     74   
Forest Hills Gardens                     42                      42     42   
Fresh Meadows                            28                      28     28   
Glen Oaks                                44                      44     44   
Glendale                                  8                       8      8   
Hammels                                  34                      34     34   
Hillcrest                                40                      40     40   
Hollis                                   28                      28     28   
Holliswood                               10                      10     10   
Howard Beach                             74                      74     74   
Hunters Point                           106                     106    106   
Jackson Heights                         162                     162    162   
Jamaica Center                           88                      88     88   
Jamaica Estates                           8                       8      8   
Jamaica Hills                            58                      58     58   
Kew Gardens                              98                      98     98   
Kew Gardens Hills                        40                      40     40   
Laurelton                                12                      12     12   
Lefrak City                              52                      52     52   
Lindenwood                               22                      22     22   
Little Neck                             106                     106    106   
Long Island City                        142                     142    142   
Malba                                     8                       8      8   
Maspeth                                  68 

In [246]:
print('There are {} distinct venues in {} categories.'.format(
    len(queens_venues['Venue'].unique()),len(queens_venues['Venue Category'].unique())))

There are 1707 distinct venues in 271 categories.


In [247]:

queens_onehot = pd.get_dummies(queens_venues[['Venue Category']], prefix="", prefix_sep="")


queens_onehot['Neighborhood'] = queens_venues['Neighborhood'] 

neighbor = queens_onehot['Neighborhood']
queens_onehot.drop(labels=['Neighborhood'], axis=1,inplace = True)
queens_onehot.insert(0, 'Neighborhood', neighbor)

queens_onehot.head()

Neighborhood  Accessories Store  Afghan Restaurant  American Restaurant  \
0      Astoria                  0                  0                    0   
1      Astoria                  0                  0                    0   
2      Astoria                  0                  0                    0   
3      Astoria                  0                  0                    0   
4      Astoria                  0                  0                    0   

   Antique Shop  Arepa Restaurant  Argentinian Restaurant  Art Gallery  \
0             0                 0                       0            0   
1             0                 0                       0            0   
2             0                 0                       0            0   
3             0                 0                       0            0   
4             0                 0                       0            0   

   Art Museum  Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0           0                    0                     0                 0   
1           0                    0                     0                 0   
2           0                    0                     0                 0   
3           0                    0                     0                 0   
4           0                    0                     0                 0   

   Athletics & Sports  Automotive Shop  BBQ Joint  Bagel Shop  Bakery  Bank  \
0                   0                0          0           0       0     0   
1                   0                0          0           0       0     0   
2                   0                0          0           0       0     0   
3                   0                0          0           0       0     0   
4                   0                0          0           0       0     0   

   Bar  Baseball Field  Basketball Court  Bath House  Beach  Beach Bar  \
0    0               0                 0           0      0          0   
1    0               0                 0           0      0          0   
2    0               0                 0           0      0          0   
3    0               0                 0           0      0          0   
4    0               0                 0           0      0          0   

   Bed & Breakfast  Beer Garden  Bike Trail  Bistro  Board Shop  Bookstore  \
0                0            0           0       0           0          0   
1                0            0           0       0           0          0   
2                0            0           0       0           0          0   
3                0            0           0       0           0          0   
4                0            0           0       0           0          0   

   Boutique  Bowling Alley  Brazilian Restaurant  Breakfast Spot  Brewery  \
0         0              0                     1               0        0   
1         0              0                     0               0        0   
2         0              0                     0               0        0   
3         0              0                     0               0        0   
4         0              0                     0               0        0   

   Bubble Tea Shop  Buffet  Building  Burger Joint  Burrito Place  Bus Line  \
0                0       0         0             0              0         0   
1                0       0         0             0              0         0   
2                0       0         0             0              0         0   
3                0       0         0             0              0         0   
4                0       0         0             0              0         0   

   Bus Station  Bus Stop  Business Service  Butcher  Café  \
0            0         0                 0        0     0   
1            0         0                 0        0     0   
2            0         0                 0        0     0   
3            0         0                 0        0   

In [249]:
queens_grouped = queens_onehot.groupby('Neighborhood').mean().reset_index()
queens_grouped

Neighborhood  Accessories Store  Afghan Restaurant  \
0                Arverne           0.000000           0.000000   
1                Astoria           0.000000           0.000000   
2        Astoria Heights           0.000000           0.000000   
3             Auburndale           0.000000           0.000000   
4            Bay Terrace           0.026316           0.000000   
5                Bayside           0.000000           0.000000   
6              Bayswater           0.000000           0.000000   
7             Beechhurst           0.000000           0.000000   
8               Bellaire           0.000000           0.000000   
9           Belle Harbor           0.000000           0.000000   
10             Bellerose           0.000000           0.000000   
11            Blissville           0.000000           0.000000   
12          Breezy Point           0.000000           0.000000   
13             Briarwood           0.000000           0.000000   
14         Broad Channel           0.000000           0.000000   
15            Brookville           0.000000           0.000000   
16       Cambria Heights           0.000000           0.000000   
17         College Point           0.000000           0.000000   
18                Corona           0.000000           0.000000   
19            Douglaston           0.000000           0.000000   
20         East Elmhurst           0.000000           0.000000   
21              Edgemere           0.000000           0.000000   
22              Elmhurst           0.000000           0.000000   
23          Far Rockaway           0.000000           0.000000   
24           Floral Park           0.000000           0.000000   
25              Flushing           0.000000           0.000000   
26          Forest Hills           0.000000           0.000000   
27  Forest Hills Gardens           0.000000           0.000000   
28         Fresh Meadows           0.000000           0.000000   
29             Glen Oaks           0.000000           0.000000   
30              Glendale           0.000000           0.000000   
31               Hammels           0.000000           0.000000   
32             Hillcrest           0.000000           0.000000   
33                Hollis           0.000000           0.000000   
34            Holliswood           0.000000           0.000000   
35          Howard Beach           0.000000           0.000000   
36         Hunters Point           0.000000           0.000000   
37       Jackson Heights           0.000000           0.000000   
38        Jamaica Center           0.000000           0.000000   
39       Jamaica Estates           0.000000           0.000000   
40         Jamaica Hills           0.000000           0.000000   
41           Kew Gardens           0.000000           0.000000   
42     Kew Gardens Hills           0.000000           0.000000   
43             Laurelton           0.000000           0.000000   
44           Lefrak City           0.000000           0.000000   
45            Lindenwood           0.000000           0.000000   
46           Little Neck           0.000000           0.000000   
47      Long Island City           0.000000           0.000000   
48                 Malba           0.000000           0.000000   
49               Maspeth           0.000000           0.000000   
50        Middle Village           0.000000           0.000000   
51           Murray Hill           0.000000           0.000000   
52              Neponsit           0.000000           0.000000   
53          North Corona           0.000000           0.000000   
54       Oakland Gardens           0.000000           0.000000   
55            Ozone Park           0.000000           0.000000   
56               Pomonok           0.000000           0.000000   
57        Queens Village           0.000000           0.000000   
58       Queensboro Hill           0.000000           0.000000   
59          Queensbridge           0.000000         

In [250]:

for hood in queens_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = queens_grouped[queens_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Arverne----
             venue  freq
0        Surf Spot  0.22
1    Metro Station  0.11
2   Sandwich Place  0.11
3  Thai Restaurant  0.06
4      Pizza Place  0.06


----Astoria----
                       venue  freq
0  Middle Eastern Restaurant  0.06
1                        Bar  0.06
2                 Hookah Bar  0.05
3         Seafood Restaurant  0.04
4   Mediterranean Restaurant  0.04


----Astoria Heights----
                venue  freq
0         Supermarket  0.08
1       Bowling Alley  0.08
2        Burger Joint  0.08
3        Cocktail Bar  0.08
4  Italian Restaurant  0.08


----Auburndale----
                venue  freq
0   Mobile Phone Shop  0.06
1               Train  0.06
2  Italian Restaurant  0.06
3            Pharmacy  0.06
4           Pet Store  0.06


----Bay Terrace----
            venue  freq
0  Clothing Store  0.11
1  Cosmetics Shop  0.05
2      Donut Shop  0.05
3  Lingerie Store  0.05
4      Shoe Store  0.05


----Bayside----
                 venue  freq
0         

In [254]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [255]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']


columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))


neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = queens_grouped['Neighborhood']

for ind in np.arange(queens_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(queens_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood            1st Most Common Venue  \
0                Arverne                        Surf Spot   
1                Astoria                              Bar   
2        Astoria Heights                            Plaza   
3             Auburndale                Korean Restaurant   
4            Bay Terrace                   Clothing Store   
5                Bayside                              Bar   
6              Bayswater                             Park   
7             Beechhurst               Chinese Restaurant   
8               Bellaire               Chinese Restaurant   
9           Belle Harbor                            Beach   
10             Bellerose               Italian Restaurant   
11            Blissville                    Deli / Bodega   
12          Breezy Point              Monument / Landmark   
13             Briarwood                Indian Restaurant   
14         Broad Channel              Sporting Goods Shop   
15            Brookville                    Deli / Bodega   
16       Cambria Heights             Caribbean Restaurant   
17         College Point                      Pizza Place   
18                Corona               Mexican Restaurant   
19            Douglaston                    Deli / Bodega   
20         East Elmhurst                       Donut Shop   
21              Edgemere                      Pizza Place   
22              Elmhurst                  Thai Restaurant   
23          Far Rockaway                    Deli / Bodega   
24           Floral Park                Indian Restaurant   
25              Flushing                Hotpot Restaurant   
26          Forest Hills             Gym / Fitness Center   
27  Forest Hills Gardens                   Sandwich Place   
28         Fresh Meadows                      Bus Station   
29             Glen Oaks                         Pharmacy   
30              Glendale                          Brewery   
31               Hammels                            Beach   
32             Hillcrest                    Grocery Store   
33                Hollis              Fried Chicken Joint   
34            Holliswood                Mobile Phone Shop   
35          Howard Beach                       Bagel Shop   
36         Hunters Point                             Café   
37       Jackson Heights        Latin American Restaurant   
38        Jamaica Center                Mobile Phone Shop   
39       Jamaica Estates                Indian Restaurant   
40         Jamaica Hills                         Pharmacy   
41           Kew Gardens               Chinese Restaurant   
42     Kew Gardens Hills                             Bank   
43             Laurelton             Caribbean Restaurant   
44           Lefrak City                 Department Store   
45            Lindenwood                    Deli / Bodega   
46           Little Neck               Chinese Restaurant   
47      Long Island City                            Hotel   
48                 Malba                        Rest Area   
49               Maspeth                      Pizza Place   
50        Middle Village                           Bakery   
51           Murray Hill                Korean Restaurant   
52              Neponsit                            Beach   
53          North Corona                    Deli / Bodega   
54       Oakland Gardens                Korean Restaurant   
55            Ozone Park                      Pizza Place   
56               Pomonok                       Playground   
57        Queens Village                             Bank   
58       Queensboro Hill               Chinese Restaurant   
59          Queensbridge                            Hotel   
60            Ravenswood               Chinese Restaurant   
61             Rego Park                           Bakery   
62         Richmond Hill                             Bank   
63             Ridgewood                             Bank   
64              Rochdale  Southern / Soul Food 

In [256]:
neighborhoods_venues_sorted.iloc[47,]

Neighborhood                Long Island City
1st Most Common Venue                  Hotel
2nd Most Common Venue            Coffee Shop
3rd Most Common Venue     Mexican Restaurant
4th Most Common Venue            Pizza Place
5th Most Common Venue                    Bar
6th Most Common Venue                   Café
7th Most Common Venue          Deli / Bodega
8th Most Common Venue                 Hostel
9th Most Common Venue             Food Truck
10th Most Common Venue                Market
Name: 47, dtype: object

Cluster Neighborhood

In [264]:
kclusters = 5

queens_grouped_clustering = queens_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(queens_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 2, 2, 1, 2, 2, 3], dtype=int32)

In [265]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

queens_merged = queens_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
queens_merged = queens_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

queens_merged.head() # check the last columns!

Borough     Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Queens          Astoria  40.768509 -73.915654               2   
1  Queens         Woodside  40.746349 -73.901842               2   
2  Queens  Jackson Heights  40.751981 -73.882821               2   
3  Queens         Elmhurst  40.744049 -73.881656               2   
4  Queens     Howard Beach  40.654225 -73.838138               2   

       1st Most Common Venue      2nd Most Common Venue  \
0                        Bar  Middle Eastern Restaurant   
1              Grocery Store            Thai Restaurant   
2  Latin American Restaurant        Peruvian Restaurant   
3            Thai Restaurant         Mexican Restaurant   
4                 Bagel Shop         Italian Restaurant   

       3rd Most Common Venue      4th Most Common Venue 5th Most Common Venue  \
0                 Hookah Bar          Indian Restaurant    Seafood Restaurant   
1        Filipino Restaurant  Latin American Restaurant                Bakery   
2  South American Restaurant          Mobile Phone Shop                Bakery   
3  South American Restaurant      Vietnamese Restaurant                  Park   
4                   Pharmacy             Clothing Store  Fast Food Restaurant   

  6th Most Common Venue     7th Most Common Venue 8th Most Common Venue  \
0         Deli / Bodega  Mediterranean Restaurant      Greek Restaurant   
1   American Restaurant               Pizza Place            Donut Shop   
2         Grocery Store        Mexican Restaurant       Thai Restaurant   
3    Chinese Restaurant                       Bar      Malay Restaurant   
4         Deli / Bodega            Sandwich Place  Gym / Fitness Center   

  9th Most Common Venue 10th Most Common Venue  
0                Bakery             Bagel Shop  
1                   Pub                    Bar  
2            Kids Store            Supermarket  
3      Sushi Restaurant   Colombian Restaurant  
4    Chinese Restaurant                    Spa

In [266]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(queens_merged['Latitude'], queens_merged['Longitude'], queens_merged['Neighborhood'], queens_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [268]:
queens_merged.loc[queens_merged['Cluster Labels'] == 0, queens_merged.columns[[1] + list(range(5, queens_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
64    Brookville         Deli / Bodega           Yoga Studio   
145   Brookville         Deli / Bodega           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
64            Flea Market                  Farm        Farmers Market   
145           Flea Market                  Farm        Farmers Market   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
64   Fast Food Restaurant   Filipino Restaurant     Fish & Chips Shop   
145  Fast Food Restaurant   Filipino Restaurant     Fish & Chips Shop   

    9th Most Common Venue 10th Most Common Venue  
64            Fish Market            Flower Shop  
145           Fish Market            Flower Shop

In [270]:
queens_merged.loc[queens_merged['Cluster Labels'] == 1, queens_merged.columns[[1] + list(range(5, queens_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
63    Somerville                  Park           Yoga Studio   
79     Bayswater                  Park            Playground   
144   Somerville                  Park           Yoga Studio   
160    Bayswater                  Park            Playground   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
63      Fish & Chips Shop    Falafel Restaurant                  Farm   
79            Yoga Studio     Fish & Chips Shop    Falafel Restaurant   
144     Fish & Chips Shop    Falafel Restaurant                  Farm   
160           Yoga Studio     Fish & Chips Shop    Falafel Restaurant   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
63         Farmers Market  Fast Food Restaurant   Filipino Restaurant   
79                   Farm        Farmers Market  Fast Food Restaurant   
144        Farmers Market  Fast Food Restaurant   Filipino Restaurant   
160                  Farm        Farmers Market  Fast Food Restaurant   

    9th Most Common Venue 10th Most Common Venue  
63            Fish Market    Empanada Restaurant  
79    Filipino Restaurant            Fish Market  
144           Fish Market    Empanada Restaurant  
160   Filipino Restaurant            Fish Market

In [271]:
queens_merged.loc[queens_merged['Cluster Labels'] == 2, queens_merged.columns[[1] + list(range(5, queens_merged.shape[1]))]]

Neighborhood            1st Most Common Venue  \
0                 Astoria                              Bar   
1                Woodside                    Grocery Store   
2         Jackson Heights        Latin American Restaurant   
3                Elmhurst                  Thai Restaurant   
4            Howard Beach                       Bagel Shop   
5                  Corona               Mexican Restaurant   
6            Forest Hills             Gym / Fitness Center   
7             Kew Gardens               Chinese Restaurant   
8           Richmond Hill                             Bank   
9                Flushing                Hotpot Restaurant   
10       Long Island City                            Hotel   
11              Sunnyside                      Pizza Place   
12          East Elmhurst                       Donut Shop   
13                Maspeth                      Pizza Place   
14              Ridgewood                             Bank   
15               Glendale                          Brewery   
16              Rego Park                           Bakery   
17              Woodhaven                    Deli / Bodega   
18             Ozone Park                      Pizza Place   
19       South Ozone Park                    Deli / Bodega   
20          College Point                      Pizza Place   
21             Whitestone                    Deli / Bodega   
22                Bayside                              Bar   
23             Auburndale                Korean Restaurant   
24            Little Neck               Chinese Restaurant   
25             Douglaston                    Deli / Bodega   
26              Glen Oaks                         Pharmacy   
27              Bellerose               Italian Restaurant   
28      Kew Gardens Hills                             Bank   
29          Fresh Meadows                      Bus Station   
30              Briarwood                Indian Restaurant   
31         Jamaica Center                Mobile Phone Shop   
32        Oakland Gardens                Korean Restaurant   
33         Queens Village                             Bank   
34                 Hollis              Fried Chicken Joint   
35          South Jamaica                      Bus Station   
36             St. Albans             Caribbean Restaurant   
37               Rochdale  Southern / Soul Food Restaurant   
38    Springfield Gardens                      Bus Station   
39        Cambria Heights             Caribbean Restaurant   
40               Rosedale                      Bus Station   
41           Far Rockaway                    Deli / Bodega   
42          Broad Channel              Sporting Goods Shop   
44               Steinway                   Cosmetics Shop   
45             Beechhurst               Chinese Restaurant   
46            Bay Terrace                   Clothing Store   
47               Edgemere                      Pizza Place   
48                Arverne                        Surf Spot   
51            Murray Hill                Korean Restaurant   
52            Floral Park                Indian Restaurant   
53             Holliswood                Mobile Phone Shop   
54        Jamaica Estates                Indian Restaurant   
55        Queensboro Hill               Chinese Restaurant   
56              Hillcrest                    Grocery Store   
57             Ravenswood               Chinese Restaurant   
58             Lindenwood                    Deli / Bodega   
59              Laurelton             Caribbean Restaurant   
60            Lefrak City                 Department Store   
65               Bellaire               Chinese Restaurant   
66           North Corona                    Deli / Bodega   
67   Forest Hills Gardens                   Sandwich Place   
68          Jamaica Hills                         Pharmacy   
69                 Utopia                    Deli / Bodega   
70                Pomonok                    

In [272]:
queens_merged.loc[queens_merged['Cluster Labels'] == 3, queens_merged.columns[[1] + list(range(5, queens_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
43     Breezy Point   Monument / Landmark                 Trail   
49   Rockaway Beach                 Beach        Ice Cream Shop   
61     Belle Harbor                 Beach         Deli / Bodega   
62    Rockaway Park                 Beach           Pizza Place   
75          Roxbury         Deli / Bodega             Irish Pub   
78          Hammels                 Beach   Fried Chicken Joint   
124    Breezy Point   Monument / Landmark                 Trail   
130  Rockaway Beach                 Beach        Ice Cream Shop   
142    Belle Harbor                 Beach         Deli / Bodega   
143   Rockaway Park                 Beach           Pizza Place   
156         Roxbury         Deli / Bodega             Irish Pub   
159         Hammels                 Beach   Fried Chicken Joint   

    3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
43                  Beach           Event Space                       Farm   
49             Food Truck      Arepa Restaurant  Latin American Restaurant   
61                    Pub                   Spa                   Boutique   
62                   Bank            Donut Shop                   Pharmacy   
75                  Beach  Fast Food Restaurant                      Trail   
78   Gym / Fitness Center                 Diner                    Dog Run   
124                 Beach           Event Space                       Farm   
130            Food Truck      Arepa Restaurant  Latin American Restaurant   
142                   Pub                   Spa                   Boutique   
143                  Bank            Donut Shop                   Pharmacy   
156                 Beach  Fast Food Restaurant                      Trail   
159  Gym / Fitness Center                 Diner                    Dog Run   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
43         Farmers Market  Fast Food Restaurant   Filipino Restaurant   
49            Bus Station           Pizza Place             BBQ Joint   
61     Mexican Restaurant    Chinese Restaurant                Bakery   
62                  Diner        Sandwich Place    Seafood Restaurant   
75         Baseball Field     Food & Drink Shop                  Food   
78   Fast Food Restaurant            Shoe Store              Bus Stop   
124        Farmers Market  Fast Food Restaurant   Filipino Restaurant   
130           Bus Station           Pizza Place             BBQ Joint   
142    Mexican Restaurant    Chinese Restaurant                Bakery   
143                 Diner        Sandwich Place    Seafood Restaurant   
156        Baseball Field     Food & Drink Shop                  Food   
159  Fast Food Restaurant            Shoe Store              Bus Stop   

    9th Most Common Venue 10th Most Common Venue  
43      Fish & Chips Shop            Fish Market  
49                  Hotel     Spanish Restaurant  
61             Bagel Shop             Donut Shop  
62             Smoke Shop      French Restaurant  
75            Flower Shop     Falafel Restaurant  
78            Bus Station             Food Truck  
124     Fish & Chips Shop            Fish Market  
130                 Hotel     Spanish Restaurant  
142            Bagel Shop             Donut Shop  
143            Smoke Shop      French Restaurant  
156           Flower Shop     Falafel Restaurant  
159           Bus Station             Food Truck

In [273]:
queens_merged.loc[queens_merged['Cluster Labels'] == 4, queens_merged.columns[[1] + list(range(5, queens_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
50      Neponsit                 Beach           Yoga Studio   
131     Neponsit                 Beach           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
50            Fish Market                  Farm        Farmers Market   
131           Fish Market                  Farm        Farmers Market   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
50   Fast Food Restaurant   Filipino Restaurant     Fish & Chips Shop   
131  Fast Food Restaurant   Filipino Restaurant     Fish & Chips Shop   

    9th Most Common Venue 10th Most Common Venue  
50            Flea Market     Frozen Yogurt Shop  
131           Flea Market     Frozen Yogurt Shop

DISCUSSION SECTION

Toronto has ten boroughs and 103 neighborhoods. Downtown Toronto borough has 1228 venues in 19 neighborhoods, and the most popular places are coffee shops and international cuisine restaurants that offer a menu for Toronto's famous financial district, which is similar to cities like New York or Chicago.

New York City has five boroughs and 612 neighborhoods. Queens borough has
found 4160 venues in 162 neighborhoods, and the most popular sites determine the culinary offer of Queens, which is the largest and most diverse district in New York. Here the city's culture is being reinvented.

CONCLUSION SECTION: CONCLUDE THE REPORT.

In conclusion, based on the number of venues and a variety of places, I would suggest considering Queens over Downtown Toronto to locate a Peruvian restaurant because Queens, offers more options for extracurricular and outdoor activities for individuals and families. Besides, Queens is New York's largest borough, in which different cultures coexist, and it is the district that is evolving fast due to art, tourism, and cinema.
